vllm/model/model.py

```diff
@@ -0,0 +1,204 @@
+import logging
+import os
+import subprocess
+import time
+import uuid
+
+import httpx
+from transformers import AutoTokenizer
+from vllm.engine.arg_utils import AsyncEngineArgs
+from vllm.engine.async_llm_engine import AsyncLLMEngine
+
+from vllm import SamplingParams
+
+os.environ["TOKENIZERS_PARALLELISM"] = "true"
+
+logger = logging.getLogger(__name__)
+
+
+class Model:
+    # TODO: better way to detect model server startup failure than using `MAX_FAILED_SECONDS`
+    MAX_FAILED_SECONDS = 600  # 10 minutes; the reason this would take this long is mostly if we download a large model
+
+    def __init__(self, **kwargs):
+        self._config = kwargs["config"]
+        self.model_id = None
+        self.llm_engine = None
+        self.model_args = None
+        self.hf_secret_token = kwargs["secrets"].get("hf_access_token", None)
+        self.openai_compatible = self._config["model_metadata"].get(
+            "openai_compatible", False
+        )
+        self.vllm_base_url = None
+        os.environ["HF_TOKEN"] = self.hf_secret_token
+
+    def load(self):
+        self._model_metadata = self._config["model_metadata"]
+        self._model_repo_id = self._model_metadata["repo_id"]
+        self._vllm_config = self._model_metadata["vllm_config"]
+        if self._vllm_config is None:
+            self._vllm_config = {}
+        logger.info(f"main model: {self._model_repo_id}")
+        logger.info(f"vllm config: {self._vllm_config}")
+        if self.openai_compatible:
+            self._client = httpx.AsyncClient(timeout=None)
+            command = ["vllm", "serve", self._model_repo_id]
+            for key, value in self._vllm_config.items():
+                if value is True:
+                    command.append(f"--{key.replace('_', '-')}")
+                elif value is False:
+                    continue
+                else:
+                    command.append(f"--{key.replace('_', '-')}")
+                    command.append(str(value))
+
+            logger.info(
+                f"Starting openai compatible vLLM server with command: {command}"
+            )
+
+            process = subprocess.Popen(
+                command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True
+            )
+
+            # Wait for 10 seconds and check if command fails
+            time.sleep(10)
```

Comment:
I might need some help understanding this flow

Reply:
Wait for 10 seconds and then check if `vllm` command is still running, to early detect any command running failure